be/src/formats/parquet/file_writer.h

```diff
@@ -67,30 +79,108 @@ class ParquetBuildHelper {
 
     static void build_compression_type(::parquet::WriterProperties::Builder& builder,
                                        const TCompressionType::type& compression_type);
+
+    static std::shared_ptr<::parquet::schema::GroupNode> make_schema(const std::vector<std::string>& file_column_names,
+                                                                     const std::vector<ExprContext*>& output_expr_ctxs);
+
+    static std::shared_ptr<::parquet::WriterProperties> make_properties(const ParquetBuilderOptions& options);
+
+private:
+    static ::parquet::schema::NodePtr _make_schema_node(const std::string& name, const TypeDescriptor& type_desc,
+                                                        ::parquet::Repetition::type rep_type);
 };
 
 class FileWriterBase {
 public:
     FileWriterBase(std::unique_ptr<WritableFile> writable_file, std::shared_ptr<::parquet::WriterProperties> properties,
                    std::shared_ptr<::parquet::schema::GroupNode> schema,
                    const std::vector<ExprContext*>& output_expr_ctxs);
+
     virtual ~FileWriterBase() = default;
 
     Status init();
+
     Status write(Chunk* chunk);
+
     std::size_t file_size() const;
+
     void set_max_row_group_size(int64_t rg_size) { _max_row_group_size = rg_size; }
+
     std::shared_ptr<::parquet::FileMetaData> metadata() const { return _file_metadata; }
+
     Status split_offsets(std::vector<int64_t>& splitOffsets) const;
+
     virtual bool closed() const = 0;
 
 protected:
     virtual void _flush_row_group() = 0;
 
 private:
-    ::parquet::RowGroupWriter* _get_rg_writer();
+    class Context {
+    public:
+        Context(int16_t max_rep_level, size_t estimated_size = 0) : _max_rep_level(max_rep_level) {
+            _idx2subcol.reserve(estimated_size);
+            _def_levels.reserve(estimated_size);
+            _rep_levels.reserve(estimated_size);
+        }
+
+        void append(int idx, int16_t def_level, int16_t rep_level) {
+            _idx2subcol.push_back(idx);
+            _def_levels.push_back(def_level);
+            _rep_levels.push_back(rep_level);
+        }
+
+        std::tuple<int, int16_t, int16_t> get(int i) const {
+            return std::make_tuple(_idx2subcol[i], _def_levels[i], _rep_levels[i]);
+        }
+
+        int size() const { return _def_levels.size(); }
+
+    public:
+        const int16_t _max_rep_level;
+        constexpr static int kNULL = -1;
+
+        std::vector<int> _idx2subcol;
+        std::vector<int16_t> _def_levels;
+        std::vector<int16_t> _rep_levels;
+    };
+
+    void _generate_rg_writer();
+
     std::size_t _get_current_rg_written_bytes() const;
 
+    Status _add_column_chunk(Context& ctx, const TypeDescriptor& type_desc, const ::parquet::schema::NodePtr& node,
```

Comment:
Extract `_add_xxx` as a new class, such as `Class DefRepLevelGenerator`

Reply:
I will wrap these methods into a new class `ChunkWriter`, which converts chunks to RDV tuples and invokes WriteBatch/WriteBatchSpaced methods. 

It's too complicated to decouple the conversion logic and WriteBatch/WriteBatchSpaced methods.

Reply:
DefRepLevel generation and writeBatch invocation can be decoupled using callback function.

Gonna refactor once the unit testing is complete.

Reply:
done